lib/Jsrt/Jsrt.cpp

```diff
@@ -2746,3 +3230,504 @@ CHAKRA_API JsRunSerializedScriptWithCallback(_In_ JsSerializedScriptLoadSourceCa
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDDetached())
+    {
+        AssertMsg(false, "Already stopped TTD.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(!scriptContext->IsTTDActive())
+    {
+        AssertMsg(false, "TTD was never started.");
+        return JsErrorCategoryUsage;
+    }
+
+    threadContext->TTDLog->PushMode(TTD::TTDMode::ExcludedExecution);
+    BEGIN_JS_RUNTIME_CALLROOT_EX(scriptContext, false)
+    {
+        threadContext->EmitTTDLogIfNeeded();
+    }
+    END_JS_RUNTIME_CALL(scriptContext);
+    threadContext->TTDLog->PopMode(TTD::TTDMode::ExcludedExecution);
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDStartTimeTravelDebugging()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+    ThreadContext* threadContext = scriptContext->GetThreadContext();
+    if(threadContext->TTDLog == nullptr)
+    {
+        AssertMsg(false, "Need to create in TTD mode.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDDetached())
+    {
+        AssertMsg(false, "Cannot re-start TTD after detach.");
+        return JsErrorCategoryUsage;
+    }
+
+    if(scriptContext->IsTTDActive())
+    {
+        AssertMsg(false, "Already started TTD.");
+        return JsErrorCategoryUsage;
+    }
+
+    scriptContext->GetThreadContext()->TTDLog->SetIntoDebuggingMode();
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDPauseTimeTravelBeforeRuntimeOperation()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+
+    if(currentContext->GetRuntime()->GetThreadContext()->TTDLog != nullptr)
+    {
+        currentContext->GetRuntime()->GetThreadContext()->TTDLog->PushMode(TTD::TTDMode::ExcludedExecution);
+    }
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDReStartTimeTravelAfterRuntimeOperation()
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    if(currentContext->GetRuntime()->GetThreadContext()->TTDLog != nullptr)
+    {
+        currentContext->GetRuntime()->GetThreadContext()->TTDLog->PopMode(TTD::TTDMode::ExcludedExecution);
+    }
+
+    return JsNoError;
+#endif
+}
+
+CHAKRA_API JsTTDNotifyHostCallbackCreatedOrCanceled(_In_ bool isCreated, _In_ bool isCancel, _In_ bool isRepeating, _In_ JsValueRef function, _In_ INT64 callbackId)
+{
+#if !ENABLE_TTD
+    return JsErrorCategoryUsage;
+#else
+    JsrtContext *currentContext = JsrtContext::GetCurrent();
+    Js::ScriptContext* scriptContext = currentContext->GetScriptContext();
+
+    if(PERFORM_JSRT_TTD_RECORD_ACTION_CHECK(scriptContext))
+    {
+        Js::JavascriptFunction* jsFunction = (function != nullptr) ? Js::JavascriptFunction::FromVar(function) : nullptr;
```

Comment:
Should have VALIDATE_INCOMING_FUNCTION check? what if this is not a JavaScript function?